### PR TITLE
[SPARK-15168][PySpark][ML] Add missing params to MultilayerPerceptronClassifier

### PR DESCRIPTION
**Pull request by holdenk**
## What changes were proposed in this pull request?

MultilayerPerceptronClassifier is missing step size, solver, and weights. Add these params. Also clarify the scaladoc a bit while we are updating these params.

Eventually we should follow up and unify the HasSolver params (filed https://issues.apache.org/jira/browse/SPARK-15169 )
## How was this patch tested?

Doc tests

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala [hunk 1]
```diff
@@ -88,7 +88,7 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
    * @group expertParam
    */
   final val weights: Param[Vector] = new Param[Vector](this, "weights",
-    " Sets the weights of the model ")
+    "Weights (either initial if before training or actual on model)")
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: Should this be consistent with Scala API doc: ```Model weights. Can be returned either after training or after explicit setting```? Meanwhile, I think we should also update the Scala param doc.

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: No I think `Model weights. Can be returned either after training or after explicit setting` leaves a lot to be understood compared to explaining that it can be the initial weights if before training or the actual weights after training on the model. I'll update the Scaladoc on the setter to match.

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: Oh wait I already updated the scaladoc on the optimizer.

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: Which param doc did you want updated?

Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: I do not have strong preference about this, I just want to ensure that the docs of Scala and Python should keep consistency.

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: So they already are the same in this PR - 

Scala:
> Weights (either initial if before training or actual on model)
Python
> Weights (either initial if before training or actual on model)

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: @yanboliang can you clarify what it is you want?

File: python/pyspark/ml/classification.py [hunk 2]
```diff
@@ -1117,6 +1137,56 @@ def getBlockSize(self):
...
+    @since("2.0.0")
+    def getSolver(self):
+        """
+        Gets the value of solver or its default value.
+        """
+        return self.getOrDefault(self.solver)
+
+    @property
+    @since("2.0.0")
+    def getOptimizer(self):
```
Comment (MLnick) [inline python/pyspark/ml/classification.py]: Just wondering, why is `getOptimizer` even there? It seems redundant to me.

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala [hunk 3]
```diff
@@ -75,8 +75,8 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
    * @group expertParam
    */
   final val solver: Param[String] = new Param[String](this, "solver",
-    " Allows setting the solver: minibatch gradient descent (gd) or l-bfgs. " +
-      " l-bfgs is the default one.",
+    "Allows setting the solver: minibatch gradient descent (gd) or l-bfgs. " +
+      "(Default l-bfgs)",
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: indentation seems off here

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala [hunk 4]
```diff
@@ -181,7 +181,7 @@ class MultilayerPerceptronClassifier @Since("1.5.0") (
   def setSeed(value: Long): this.type = set(seed, value)
 
   /**
-   * Sets the model weights.
+   * Sets the initial weights used for the optimizer.
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: Perhaps "Sets the weight vector used to initialize training" might be clearer? Though I don't feel too strongly about it.

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala (hunk 1 above)
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: Perhaps we can clarify: `Model weight vector. If set before training, this represents the weight vector used to initialize training. If set on a trained model, this explicitly sets the model weights (overriding any existing weights)`, or something along those lines. Scala doc could be updated accordingly (agree it is unclear currently).

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala (hunk 3 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: So you mentioned something similar in another PR - but this seems to be the standard we follow (both in the file and others) for multi line strings with + which is add 2 for original indent for multi-line and then add two for + indent. I'm not saying this a good style - but it does seem to be the one we follow.

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala (hunk 1 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: Sure I'm down with updating both :)

File: python/pyspark/ml/classification.py (hunk 2 above)
Comment (holdenk) [inline python/pyspark/ml/classification.py]: It is somewhat of a duplicate but I just added it to match the Scala API. I'm pretty ambivalent on the actual method its self.

Comment (MLnick) [inline python/pyspark/ml/classification.py]: Sorry want clear - yeah agree it should match Scala. But maybe we should
just deprecate it, seems superfluous.

Comment (holdenk) [inline python/pyspark/ml/classification.py]: Sounds legit, ok to add the depercation in this PR?

Comment (MLnick) [inline python/pyspark/ml/classification.py]: @avulanov can you weigh in on this?

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala (hunk 3 above)
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: I've seen both styles. It's not clear which is the "accepted one". I think it's this way because that's what the IDE does. Let's leave it as is though.

File: python/pyspark/ml/classification.py (hunk 2 above)
Comment (yanboliang) [inline python/pyspark/ml/classification.py]: I suspect that it should be ```getSolver```. I found there are some issues should be fixed at Scala side and send #13076, please feel free to comment. After that, we can fix the issues at Python side by this PR.

File: python/pyspark/ml/classification.py [hunk 5]
```diff
@@ -1058,6 +1059,14 @@ class MultilayerPerceptronClassifier(JavaEstimator, HasFeaturesCol, HasLabelCol,
     True
     >>> model.weights == model2.weights
     True
+    >>> mlp2 = mlp2.setWeights([
+    ...    2, 5, 1, -7, -5, -10, 0, 0.6, -1, 2, -2, 1, 2, -7, -1, -2, 2, 1, -1, 9, -9, 3, -3, -3,
+    ...    3.0, 0, -1])
```
Comment (yanboliang) [inline python/pyspark/ml/classification.py]: Minor: Organize this vector into two line.

File: python/pyspark/ml/classification.py (hunk 2 above)
Comment (avulanov) [inline python/pyspark/ml/classification.py]: Thank you for bringing that issue, @MLnick. `getOptimizer` is there to let the user know explicitly which optimizer is used. The reason why I did not implement `setOpmizer` is because minibatch gradient descent in not very efficient in Spark and can be used only in very specific cases. So the default choice for Spark is L-BFGS. However, if we label `setOpmizer` as `expertParam` then it is OK to have it. I believe that we should proceed as suggested by @yanboliang

Comment (MLnick) [inline python/pyspark/ml/classification.py]: Sorry for some reason I had thought we had both `getSolver` and `getOptimizer`. I see we only have `getOptimizer`. I agree we should rename that to `getSolver` instead.

File: python/pyspark/ml/classification.py [hunk 6]
```diff
@@ -1063,6 +1064,13 @@ class MultilayerPerceptronClassifier(JavaEstimator, HasFeaturesCol, HasLabelCol,
     True
     >>> model.weights == model2.weights
     True
+    >>> mlp2 = mlp2.setInitialWeights([2, 5, 1, -7, -5, -10, 0, 0.6, -1, 2, -2, 1, 2, -7, -1, -2,
+    ...    2, 1, -1, 9, -9, 3, -3, -3, 3.0, 0, -1])
```
Comment (yanboliang) [inline python/pyspark/ml/classification.py]: I think we can change the values of ```layers``` to make the model and ```setInitialWeights``` more succinct in doctest.

Comment (holdenk) [inline python/pyspark/ml/classification.py]: We could, but this way its matching the tests in Scala

Comment (yanboliang) [inline python/pyspark/ml/classification.py]: In MLlib, Scala unit tests focus on verifying the correctness, Java unit tests focus on compatibility verification, and Python doc tests should focus on providing examples to users. So I think we should keep Python doc tests succinct just as you see in the code base.

Comment (yanboliang) [inline python/pyspark/ml/classification.py]: After we simplify the model, we can directly test the value of ```model.weights```.

**Comment by holdenk** [comment]
So I simplified the test down a fair amount, didn't switch to printing the model weights since that seems like it could be flaky with floats (I can of course use ... in doctests if we want but I don't think it adds much).

**Comment by yanboliang** [comment]
I don't have strong preference on printing the model weights issues. This looks good for me. cc @MLnick @jkbradley for another pass.

**Comment by holdenk** [comment]
ping @MLnick ?

**Comment by holdenk** [comment]
ping?

**Comment by MLnick** [comment]
jenkins retest this please

**Comment by MLnick** [comment]
Merged to master/branch-2.0. Thanks!
